### DISCUSSION TITLE
Diode

### DISCUSSION THREAD

**Discussion by edsonacordi** [logisim-evolution/logisim-evolution#808] [Ideas]
It would be nice if it were possible to use diodes to implement logical ports.

**Comment by maehne** [comment]
Do I understand you correctly that you would like to illustrate the functioning of simple diode logic gates as, e.g., described in [this Wikipedia article](https://en.wikipedia.org/wiki/Diode_logic)? You can fake already this behaviour using a PMOS with its gate connected to GND for the diode connecting the input to the pull-down resistor and an NMOS with its gate connected to VDD for the diode connecting the input to the pull-up resistor. You can hide this by creating a subcircuit for both diode connection cases: 

![diode_logic](https://user-images.githubusercontent.com/1174243/72621978-8fb74980-3942-11ea-9f0e-83165f4d3489.png)

The ZIP-archive attached to this message contains a circuit file illustrating this approach. However, be aware that this is a quite crude model. I would rather illustrate the functioning of the different logic families using schematics for a SPICE simulator. E.g., LTspice is a suitable tool for this.

Based on the Java code already implementing the MOSFET primitive in Logisim, it should be relatively easy to implement this kind of diode model in Logisim. A PR is welcome.

[diode_logic.zip](https://github.com/reds-heig/logisim-evolution/files/4077334/diode_logic.zip)

**Reply by mbaillif** [reply]
That's exactly right. Top.

At the beginning we were talking about removing transistors from Logisim, clearly insufficient if we think about electronics but presentable when we talk about logic.

Logisim-evolution is clearly a logic simulator and we are not going to compare it to Ltspice, Qucs or others.

the question is : 
We keep the transistors and possibly add diodes as components. Of course it is necessary in the documentation to issue the recommendations of use.

Or

We remove the transistors and we can also present in the documentation tips and tricks to create them in subcircuits.

Personally I have a soft spot for these basic components, more to present the history of logic than for an electronics course.
This century there are few people building logic circuits from CD4007 :-)

**Reply by maehne** [reply]
If we keep the transistors and properly draw attention to their limitations in the documentation, I am not opposed to adding a dedicated diode component. Its purpose, function, and limitations need just be documented in a similar fashion.

**Reply by mbaillif** [reply]
I am satisfied with the sub-circuit solution.
And indeed diodes are analog components.

**Comment by edsonacordi** [comment]
Hello maehne, thank you for your reply.
Indeed, using a subcircuit to hide the transistor, your solution is satisfactory. I use LTspice and Qucs for simulations involving analog electronics. Thanks for sharing!

**Comment by Anikyte** [comment]
Surely a [buffer](https://i.imgur.com/ZS7YbLy.png) would do the trick? As if I understand it correctly, you're using it for one way flow, and it's much smaller than the above solution.

**Comment by maehne** [comment]
Feel free to provide a circuit demonstrating how you would use a buffer to model the above diode logic gates. The tricky thing is that the output of the diode needs to be either '1' or 'Z' for the OR case or '0' or 'Z' for the AND case. I don't see a way to achieve this behaviour with a buffer.

**Comment by IArnaut2** [comment]
I find the lack of diodes baffling. I need diodes for a college homework where I must make a circuit that decodes three diodes.

**Reply by maehne** [reply]
Please checkout the workaround above and explanation why we are not eager to add diodes to Logisim-evolution as components. Like for transistors, one would need to be very careful how to use them. Keeping their model’s inherent limitations in mind and using them accordingly in the right circuit topologies to achieve a behaviour matching reality. You are better of to design and model transistor-level logic circuits in a SPICE simulator. Nevertheless, we are open to PRs trying to improve Logisim-evolution’s capabilities in this domain as long as they don’t compromise its main purpose.
